### PR TITLE
add hint-anchors for extension health-checks

### PR DESCRIPTION
**Pull request by Pothulapati**
This PR adds hint-anchors for the viz and jaeger extensions

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

### CODE REVIEW

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 1]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+
+```bash
+$ kubectl -n linkerd-viz get pods
+NAME                                   READY   STATUS    RESTARTS   AGE
+linkerd-grafana-68cddd7cc8-nrv4h       2/2     Running   3          18h
+linkerd-metrics-api-77f684f7c7-hnw8r   2/2     Running   2          18h
+linkerd-prometheus-5f6898ff8b-s6rjc    2/2     Running   2          18h
+linkerd-tap-59f5595fc7-ttndp           2/2     Running   2          18h
+linkerd-web-78d6588d4-pn299            2/2     Running   2          18h
+tap-injector-566f7ff8df-vpcwc          2/2     Running   2          18h
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1945]: :point_up:

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 2]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+Ensure all the linkerd-viz pods are running with 2/2
+
+```bash
+$ kubectl -n linkerd-viz get pods
+NAME                                   READY   STATUS    RESTARTS   AGE
+linkerd-grafana-68cddd7cc8-nrv4h       2/2     Running   3          18h
+linkerd-metrics-api-77f684f7c7-hnw8r   2/2     Running   2          18h
+linkerd-prometheus-5f6898ff8b-s6rjc    2/2     Running   2          18h
+linkerd-tap-59f5595fc7-ttndp           2/2     Running   2          18h
+linkerd-web-78d6588d4-pn299            2/2     Running   2          18h
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1968]: :point_up:

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 3]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+```
+
+Make sure that the `proxy-injector` is working correctly by running
+`linkerd check`
+
+### √ prometheus is installed and configured correctly {#l5d-viz-prometheus}
+
+```bash
+× prometheus is installed and configured correctly
+    missing ClusterRoles: linkerd-linkerd-viz-prometheus
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1979]:
```suggestion
    missing ClusterRoles: linkerd-viz-prometheus
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 4]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+correctly.
+
+```bash
+❯ k -n linkerd-viz get deploy,cm | grep prometheus
+deployment.apps/prometheus     1/1     1            1           3m18s
+configmap/prometheus-config   1      3m18s
+❯ k get clusterRoleBindings | grep prometheus
+linkerd-linkerd-viz-prometheus                         ClusterRole/linkerd-linkerd-viz-prometheus                         3m37s
+❯ k get clusterRoles | grep prometheus
+linkerd-linkerd-viz-prometheus                                         2021-02-26T06:03:11Zh
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1993]: :point_up:

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 5]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+```
+
+### √ can initialize the client {#l5d-viz-existence-client}
+
+Example failure:
+
+```bash
+× can initialize the client
+    Failed to get deploy for pod linkerd-metrics-api-77f684f7c7-hnw8r: not running
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:2002]:
```suggestion
    Failed to get deploy for pod metrics-api-77f684f7c7-hnw8r: not running
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 6]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+
+```bash
+× viz extension self-check
+    No results returned
+```
+
+Check the logs on the viz extensions's metrics API:
+
+```bash
+kubectl -n linkerd-viz logs deploy/linkerd-metrics-api metrics-api
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:2031]: :point_up:

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 7]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+```
+
+## The "linkerd-jaeger" checks {#l5d-jaeger}
+
+These checks only run when the `linkerd-jaeger` extension is installed. 
+This flag is intended to verify the installation of linkerd-jaeger
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:2037]: check?

**Review by alpeb — COMMENTED** [review]
Can you move the sections: `l5d-tap-cert-valid` and `l5d-tap-api` to `l5d-viz` (the latter needs the secret name to be updated and the former also requires some updates)?

Also, the sections for `l5d-jaeger-collector-running`, `l5d-jaeger-jaeger-running` and `l5d-jaeger-data-plane-trace` are missing.

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 8]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+A full list of configurable values can be found at
+https://www.github.com/linkerd/linkerd2/tree/main/viz/charts/linkerd-viz/README.md
+
+### √ linkerd-viz ClusterRoles exist {#l5d-viz-cr-exists}
+
+Example failure:
+
+```bash
+× linkerd-viz ClusterRoles exist
+    missing ClusterRoles: linkerd-linkerd-viz-metrics-api
```
[the remark this answers is not in the document]
Comment (Pothulapati) [reply linkerd.io/content/2/tasks/troubleshooting.md:1876]: They are still marked as this in the latest edges
```
❯ k get clusterRoles | grep linkerd-vi
linkerd-linkerd-viz-metrics-api                                        2021-03-01T08:25:56Z
linkerd-linkerd-viz-prometheus                                         2021-03-01T08:25:57Z
linkerd-linkerd-viz-tap                                                2021-03-01T08:25:57Z
linkerd-linkerd-viz-tap-admin                                          2021-03-01T08:25:57Z
linkerd-linkerd-viz-web-api                                            2021-03-01T08:25:58Z
linkerd-linkerd-viz-web-check                                          2021-03-01T08:25:57Z
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 9]
```diff
@@ -1842,3 +1842,308 @@ respectively.
...
+
+Ensure the linkerd-viz extension ClusterRoles exist:
+
+```bash
+$ kubectl get clusterroles | grep linkerd-viz
+linkerd-linkerd-viz-metrics-api                                        2021-01-26T18:02:17Z
+linkerd-linkerd-viz-prometheus                                         2021-01-26T18:02:17Z
+linkerd-linkerd-viz-tap                                                2021-01-26T18:02:17Z
+linkerd-linkerd-viz-tap-admin                                          2021-01-26T18:02:17Z
+linkerd-linkerd-viz-web-check                                          2021-01-26T18:02:18Z
```
[the remark this answers is not in the document]
Comment (Pothulapati) [reply linkerd.io/content/2/tasks/troubleshooting.md:1888]:
```
tarun in dev in on  kind-kind () website on  tararun/extensions-hint-anchorss [!]
❯ k get clusterRoles | grep linkerd-vi
linkerd-linkerd-viz-metrics-api                                        2021-03-01T08:25:56Z
linkerd-linkerd-viz-prometheus                                         2021-03-01T08:25:57Z
linkerd-linkerd-viz-tap                                                2021-03-01T08:25:57Z
linkerd-linkerd-viz-tap-admin                                          2021-03-01T08:25:57Z
linkerd-linkerd-viz-web-api                                            2021-03-01T08:25:58Z
linkerd-linkerd-viz-web-check                                          2021-03-01T08:25:57Z

```

still seems to be present 🤔

**Comment by Pothulapati** [comment]
>Also, the sections for l5d-jaeger-collector-running, l5d-jaeger-jaeger-running and l5d-jaeger-data-plane-trace are missing.

They were being merged into a single `l5d-jaeger-running` check as part of https://github.com/linkerd/linkerd2/pull/5673. As its under hold, I will raise another PR exclusively for this change in the linkerd2 repo.

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 10]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+yes
+```
+
+### √ tap API server has valid cert {#l5d-tap-cert-valid}
+
+Example failure:
+
+```bash
+× tap API server has valid cert
+    secrets "linkerd-tap-tls" not found
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1904]:
```suggestion
    secrets "tap-k8s-tls" not found
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 11]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+    see https://linkerd.io/checks/#l5d-tap-cert-valid for hints
+```
+
+Ensure that the `linkerd-tap-k8s-tls` secret exists and contains the appropriate
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1908]:
```suggestion
Ensure that the `tap-k8s-tls` secret exists and contains the appropriate
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 12]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+`tls.crt` and `tls.key` data entries. For versions before 2.9, the secret is
+named `linkerd-tap-tls` and it should contain the `crt.pem` and `key.pem` data
+entries.
+
+```bash
+× tap API server has valid cert
+    cert is not issued by the trust anchor: x509: certificate is valid for xxxxxx, not linkerd-tap.linkerd-viz.svc
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1915]:
```suggestion
    cert is not issued by the trust anchor: x509: certificate is valid for xxxxxx, not tap.linkerd-viz.svc
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 13]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+    see https://linkerd.io/checks/#l5d-tap-cert-valid for hints
+```
+
+Here you need to make sure the certificate was issued specifically for
+`linkerd-tap.linkerd-viz.svc`.
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1920]:
```suggestion
`tap.linkerd-viz.svc`.
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 14]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+```bash
+$ kubectl -n linkerd-jaeger get configmap/collector-config
+NAME             DATA   AGE
+collector-config   1      61m
+```
+
+Also ensure you have permission to create ConfigMaps:
+
+```bash
+$ kubectl -n linker-jaeger auth can-i create configmap
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:2133]:
```suggestion
$ kubectl -n linkerd-jaeger auth can-i create configmap
```

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 15]
```diff
@@ -1842,3 +1813,365 @@ respectively.
...
+```
+
+Make sure that the `proxy-injector` is working correctly by running
+`linkerd check`
+
+### √ jaeger extension pods are running {#l5d-jaeger-pods-running}
+
+```bash
+× jaeger extension pods are running
+    container linkerd-proxy in pod linkerd-jaeger-59f5595fc7-ttndp is not ready
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:2162]:
```suggestion
    container linkerd-proxy in pod jaeger-59f5595fc7-ttndp is not ready
```

**Review by alpeb — COMMENTED** [review]
Thanks for the follow-up :+1: 

While grepping into all the anchors found in the linkerd2 code, I couldn't find here:

- `extensions` anchor referred to in `cli/cmd/check.go`
- Anchors in `web/srv/testdata/api_check_output.json` (those have been orphaned probably for a while)
- `l5d-api-control-client` in `healthcheck.go` (probably stemming from my #5848)
- Many `l5d-multicluster-*` anchors from `multicluster/cmd/check.go` missing.

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 16]
```diff
@@ -1166,6 +1166,15 @@ Example warning:
 This happens when one of the control plane pods doesn't have at least two
 replicas running. This is likely caused by insufficient node resources.
 
+### The "extensions" checks {#extensions}
+
+Example error:
+
+```bash
+invalid extension check output from \"%s\" (JSON object expected):\n%s\n[%s]
+```
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1113]: For this particular example, we might say something about the extension binary needing to properly implement the `--output json` flag.
Then another example of error in this case is `Linkerd command xxx exists` failing, in which case we can say the command can be located on the $PATH, and provide a link to the extensions docs for more info.

File: linkerd.io/content/2/tasks/troubleshooting.md [hunk 17]
```diff
@@ -1406,164 +1415,94 @@ Done configuring CNI. Sleep=true
...
+× Link CRD exists
+    multicluster.linkerd.io/Link CRD is missing
+    see https://linkerd.io/checks/#l5d-multicluster-link-crd-exists for hints
 ```
 
-Note, it takes a little bit for pods to be scheduled, images to be pulled and
-everything to start up. If this is a permanent error, you'll want to validate
-the state of the controller pod with:
+Make sure multicluster extension is correctly installed and that the 
+`linkerd.multicluster.linkerd.io` CRD is present.
```
Comment (alpeb) [inline linkerd.io/content/2/tasks/troubleshooting.md:1372]:
```suggestion
`links.multicluster.linkerd.io` CRD is present.
```

**Comment by alpeb** [comment]
> * Anchors in `web/srv/testdata/api_check_output.json` (those have been orphaned probably for a while)

Never mind this; I just realized it's just test data...

**Review by alpeb — APPROVED** [review]
Awesome :+1: , thanks @Pothulapati !

**Pull request merged**